tools/ratbag-command.c

```diff
@@ -1435,6 +1455,292 @@ static const struct ratbag_cmd cmd_button = {
 };
 
 static int
+ratbag_cmd_led_set_bright(const struct ratbag_cmd *cmd,
+    struct ratbag *ratbag,
+    struct ratbag_cmd_options *options,
+    int argc, char **argv) {
+	struct ratbag_device *device;
+	struct ratbag_led *led;
+	char* str, *endptr;
+	int bright, b;
+	int rc;
+
+	if (argc < 1)
+		return ERR_USAGE;
+
+	led = options->led;
+	str = argv[0];
+	bright = strtoul(str, &endptr, 10);
+	if (*endptr != '\0')
+		return ERR_USAGE;
+
+	device = options->device;
+
+	rc = ratbag_led_set_bright(led, bright);
+	if (rc != 0)
+		return ERR_DEVICE;
+
+	return SUCCESS;
+}
+
+static const struct ratbag_cmd cmd_led_set_bright = {
+	.name = "bright",
+	.cmd = ratbag_cmd_led_set_bright,
+	.flags = FLAG_NEED_DEVICE | FLAG_NEED_PROFILE | FLAG_NEED_LED,
+	.subcommands = {
+		NULL,
+	},
+};
+
+static int
+ratbag_cmd_led_set_rate(const struct ratbag_cmd *cmd,
+    struct ratbag *ratbag,
+    struct ratbag_cmd_options *options,
+    int argc, char **argv) {
+
+	struct ratbag_device *device;
+	struct ratbag_led *led;
+	char* str, *endptr;
+	int rate, b;
+	int rc;
+
+	if (argc < 1)
+		return ERR_USAGE;
+
+	led = options->led;
+	str = argv[0];
+	rate = strtoul(str, &endptr, 10);
+	if (*endptr != '\0')
+		return ERR_USAGE;
+
+	rc = ratbag_led_set_rate(led, rate);
+	if (rc != 0)
+		return ERR_DEVICE;
+
+	return SUCCESS;
+}
+
+static const struct ratbag_cmd cmd_led_set_rate = {
+	.name = "rate",
+	.cmd = ratbag_cmd_led_set_rate,
+	.flags = FLAG_NEED_DEVICE | FLAG_NEED_PROFILE | FLAG_NEED_LED,
+	.subcommands = {
+		NULL,
+	},
+};
+
+static int
+ratbag_cmd_led_set_color(const struct ratbag_cmd *cmd,
+    struct ratbag *ratbag,
+    struct ratbag_cmd_options *options,
+    int argc, char **argv) {
+	struct ratbag_device *device;
+	struct ratbag_led *led;
+	char* str, *endptr;
+	int rc;
+
+	if (argc < 1)
+		return ERR_USAGE;
+
+  led = options->led;
+	str = argv[0];
+
+	struct ratbag_color color;
+	char *p;
+	char red[2], green[2], blue[2];
+
+	memcpy(&red, &str[0], 2);
+	memcpy(&green, &str[2], 2);
+	memcpy(&blue, &str[4], 2);
```

Comment:
whoah, no. please use sscanf  here which makes the whole lot a lot easier. especially given that you have an out-of-bounds writ in the line directly below :)